src/topology/sheaves/sheaf.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import topology.sheaves.presheaf
+import category_theory.limits.punit
+import category_theory.limits.shapes.products
+import category_theory.limits.shapes.equalizers
+import category_theory.full_subcategory
+
+/-!
+# Sheaves
+
+We define sheaves on a topological space, with values in an arbitrary category.
+
+The sheaf condition for a `F : presheaf C X` requires that the morphism
+`F.obj U ⟶ ∏ F.obj (U i)` (where `U` is some open set which is the union of the `U i`)
+is the equalizer of the two morphisms
+`∏ F.obj (U i) ⟶ ∏ F.obj (U i) ⊓ (U j)`.
+
+We provide the instance `category (sheaf C X)` as the full subcategory of presheaves,
+and the fully faithful functor `sheaf.forget : sheaf C X ⥤ presheaf C X`.
+-/
+
+universes v u
+
+open category_theory
+open category_theory.limits
+open topological_space
+open opposite
+open topological_space.opens
+
+namespace Top
+
+variables {C : Type u} [category.{v} C] [has_products C]
+variables {X : Top.{v}} (F : presheaf C X) {ι : Type v} (U : ι → opens X)
+
+namespace sheaf_condition
+
+/-- The product of the sections of a presheaf over a family of open sets. -/
+def pi_opens : C := ∏ (λ i : ι, F.obj (op (U i)))
+/--
+The product of the sections of a presheaf over the pairwise intersections of
+a family of open sets.
+-/
+def pi_inters : C := ∏ (λ p : ι × ι, F.obj (op (U p.1 ⊓ U p.2)))
+
+/--
+The morphism `Π F.obj (U i) ⟶ Π F.obj (U i) ⊓ (U j)` whose components
+are given by the restriction maps from `U i` to `U i ⊓ U j`.
+-/
+def left_res : pi_opens F U ⟶ pi_inters F U :=
+pi.lift (λ p : ι × ι, pi.π _ p.1 ≫ F.map (inf_le_left (U p.1) (U p.2)).op)
+
+/--
+The morphism `Π F.obj (U i) ⟶ Π F.obj (U i) ⊓ (U j)` whose components
+are given by the restriction maps from `U j` to `U i ⊓ U j`.
+-/
+def right_res : pi_opens F U ⟶ pi_inters F U :=
+pi.lift (λ p : ι × ι, pi.π _ p.2 ≫ F.map (inf_le_right (U p.1) (U p.2)).op)
+
+/--
+The morphism `F.obj U ⟶ Π F.obj (U i)` whose components
+are given by the restriction maps from `U j` to `U i ⊓ U j`.
+-/
+def res : F.obj (op (supr U)) ⟶ pi_opens F U :=
+pi.lift (λ i : ι, F.map (topological_space.opens.le_supr U i).op)
+
+lemma w : res F U ≫ left_res F U = res F U ≫ right_res F U :=
+begin
+  dsimp [res, left_res, right_res],
+  ext,
+  simp only [limit.lift_π, limit.lift_π_assoc, fan.mk_π_app, category.assoc],
+  rw [←F.map_comp],
+  rw [←F.map_comp],
+  congr,
+end
+
+/--
+The equalizer diagram for the sheaf condition.
+-/
+@[reducible]
```

Comment:
Is there a reason you use `reducible` here over `abbreviation`?

Reply:
As it seems some people are suspicious of `abbreviation`, I try to use `@[reducible]` where it works. In fact, I think if anything I should make an attempt to go through mathlib replacing `abbreviation` with `@[reducible]` as much as possible, in order to properly understand the boundary. I think the point is just that the behaviour of `@[reducible]` is better, and more widely, understood.